File: Backend-Folder/server/routes/telemetry.ts
import { Router } from 'express';
import { db } from '../db';

const router = Router();

// GET /api/telemetry/inspectors - Active field officers
router.get('/inspectors', (req, res) => {
  try {
    const list = db.getTelemetry();
    res.json({ success: true, count: list.length, data: list });
  } catch (err: any) {
    res.status(500).json({ success: false, error: err.message });
  }
});

// POST /api/telemetry/ping - Real-time location ping & geofence verification
router.post('/ping', (req, res) => {
  try {
    const { inspectorId, lat, lng, accuracyMeters } = req.body;
    if (!inspectorId || lat === undefined || lng === undefined) {
      return res.status(400).json({
        success: false,
        error: 'inspectorId, lat, and lng are required',
      });
    }

    const updated = db.pingTelemetry(
      inspectorId,
      Number(lat),
      Number(lng),
      accuracyMeters ? Number(accuracyMeters) : undefined
    );

    if (!updated) {
      return res.status(404).json({ success: false, error: 'Inspector not found' });
    }

    res.json({
      success: true,
      message: 'Telemetry updated',
      data: updated,
    });
  } catch (err: any) {
    res.status(400).json({ success: false, error: err.message });
  }
});

export default router;

File: Backend-Folder/backend/app/routes/telemetry.py
from fastapi import APIRouter, Depends, HTTPException, Request
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.services.telemetry_service import telemetry_service

router = APIRouter(prefix="/telemetry", tags=["Telemetry"])

@router.get("/inspectors")
@router.get("", include_in_schema=False)
@router.get("/", include_in_schema=False)
def get_inspectors(db: Session = Depends(get_db)):
    list_data = telemetry_service.get_inspectors(db)
    return {"success": True, "count": len(list_data), "data": list_data}

@router.post("/ping")
async def ping_telemetry(request: Request, db: Session = Depends(get_db)):
    try:
        body = await request.json()
        inspector_id = body.get("inspectorId")
        lat = body.get("lat") if body.get("lat") is not None else body.get("latitude")
        lng = body.get("lng") if body.get("lng") is not None else body.get("longitude")
        accuracy_meters = body.get("accuracyMeters")

        if not inspector_id or lat is None or lng is None:
            raise HTTPException(status_code=400, detail="inspectorId, lat, and lng are required")

        updated = telemetry_service.ping_telemetry(
            db=db,
            inspector_id=inspector_id,
            lat=float(lat),
            lng=float(lng),
            accuracy_meters=float(accuracy_meters) if accuracy_meters is not None else 5.0,
        )

        if not updated:
            raise HTTPException(status_code=404, detail="Inspector not found")

        return {"success": True, "message": "Telemetry updated", "data": updated}
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))
